src/scope.ml

```diff
@@ -42,39 +42,44 @@ let open_sign : sig_state -> Sign.t -> sig_state = fun ss sign ->
-        let txt = if b then " or variable" else "" in
-        fatal pos "Unbound symbol%s [%s]." txt s
-      end
-  | [m] when StrMap.mem m st.aliases -> (* Aliased module path. *)
-      begin
-        (* The signature must be loaded (alias is mapped). *)
-        let sign =
-          try PathMap.find (StrMap.find m st.aliases) Timed.(!Sign.loaded)
-          with _ -> assert false (* cannot fail. *)
-        in
-        (* Look for the symbol. *)
-        try (Sign.find sign s, Alias m) with Not_found ->
-        fatal pos "Unbound symbol [%a.%s]." Files.pp_path mp s
-      end
-  | _                                -> (* Fully-qualified symbol. *)
-      begin
-        (* Check that the signature was required (or is the current one). *)
-        if mp <> st.signature.sign_path then
-          if not (PathMap.mem mp !(st.signature.sign_deps)) then
-            fatal pos "No module [%a] required." Files.pp_path mp;
-        (* The signature must have been loaded. *)
-        let sign =
-          try PathMap.find mp Timed.(!Sign.loaded)
-          with Not_found -> assert false (* Should not happen. *)
-        in
-        (* Look for the symbol. *)
-        try (Sign.find sign s, Qualified) with Not_found ->
-        fatal pos "Unbound symbol [%a.%s]." Files.pp_path mp s
-      end
+  let s, pp_h =
+    match mp with
+    | []                               -> (* Symbol in scope. *)
+        begin
+          try (fst (StrMap.find s st.in_scope), Nothing) with Not_found ->
+          let txt = if b then " or variable" else "" in
+          fatal pos "Unbound symbol%s [%s]." txt s
+        end
+    | [m] when StrMap.mem m st.aliases -> (* Aliased module path. *)
+        begin
+          (* The signature must be loaded (alias is mapped). *)
+          let sign =
+            try PathMap.find (StrMap.find m st.aliases) Timed.(!Sign.loaded)
+            with _ -> assert false (* cannot fail. *)
+          in
+          (* Look for the symbol. *)
+          try (Sign.find sign s, Alias m) with Not_found ->
+          fatal pos "Unbound symbol [%a.%s]." Files.pp_path mp s
+        end
+    | _                                -> (* Fully-qualified symbol. *)
+        begin
+          (* Check that the signature was required (or is the current one). *)
+          if mp <> st.signature.sign_path then
+            if not (PathMap.mem mp !(st.signature.sign_deps)) then
+              fatal pos "No module [%a] required." Files.pp_path mp;
+          (* The signature must have been loaded. *)
+          let sign =
+            try PathMap.find mp Timed.(!Sign.loaded)
+            with Not_found -> assert false (* Should not happen. *)
+          in
+          (* Look for the symbol. *)
+          try (Sign.find sign s, Qualified) with Not_found ->
+          fatal pos "Unbound symbol [%a.%s]." Files.pp_path mp s
+        end
+  in
+  if s.sym_expo = Local && s.sym_path <> st.signature.sign_path
+  then fatal pos "Foreign local symbol used";
```

Comment:
This case cannot happen as compile filters local symbols out.

Reply:
Compile only concerns signatures written to disk, doesn't it?  We need this line to satisfy the test `tests/KO/localsymb.lp`.

Reply:
That's not nice... We need to change that. I think that we should slightly modify https://github.com/Deducteam/lambdapi/blob/1f6ec9864a08110925d5fa2a7f22ed4067fd6c63/src/compile.ml#L52-L67 . After https://github.com/Deducteam/lambdapi/blob/1f6ec9864a08110925d5fa2a7f22ed4067fd6c63/src/compile.ml#L64 and before https://github.com/Deducteam/lambdapi/blob/1f6ec9864a08110925d5fa2a7f22ed4067fd6c63/src/compile.ml#L65, we should filter out of sign the local symbols.

Reply:
Indeed, it is a way better idea